### PR TITLE
don't delete KUBE-MARK-MASQ chain in iptables/ipvs proxier

### LINKED ISSUE #68447: iptables-restore: line 7 failed in kube-proxy
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug

/sig network

**What happened**:

Kube-proxy (k8s.gcr.io/kube-proxy-amd64:v1.11.1) deployed as daemonset with ipvs mode produce during startup following error:
```
E0905 20:09:11.733419       1 proxier.go:423] Failed to execute iptables-restore for nat: exit status 1 (iptables-restore: line 7 failed
)
```

With verbosity -v5 the logs looks like this:
```
I0905 20:09:11.677674       1 iptables.go:323] running iptables-save [-t nat]
I0905 20:09:11.731479       1 iptables.go:383] running iptables-restore [-T nat --noflush --counters]
E0905 20:09:11.733419       1 proxier.go:423] Failed to execute iptables-restore for nat: exit status 1 (iptables-restore: line 7 failed
)
I0905 20:09:11.733681       1 iptables.go:323] running iptables-save [-t filter]
I0905 20:09:11.737057       1 iptables.go:383] running iptables-restore [-T filter --noflush --counters]
I0905 20:09:11.738971       1 server.go:448] Version: v1.11.1
```

Looks like proxier is trying to perform cleanup part: https://github.com/kubernetes/kubernetes/blob/v1.11.1/pkg/proxy/iptables/proxier.go#L423 but failed. Such error is generated only once during startup.

**What you expected to happen**:

Error should not exist :)

**Environment**:
- Kubernetes version (use `kubectl version`): 1.11.1
- Cloud provider or hardware configuration: none
- OS (e.g. from /etc/os-release): Ubuntu 16.04 LTS
- Kernel (e.g. `uname -a`):
- Install tools: custom script based on docker
- Others:

### PR DESCRIPTION
**Pull request by liuxu623**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
https://github.com/kubernetes/kubernetes/issues/68447#issuecomment-514195313
```
kube-proxy --cleanup

E0905 20:09:11.733419       1 proxier.go:423] Failed to execute iptables-restore for nat: exit status 1 (iptables-restore: line 7 failed
)
```
When delete KUBE-MARK-MASQ chain in iptables proxier, it linked by ipvs chains KUBE-LOAD-BALANCER and KUBE-NODE-PORT, we shouldn't delete KUBE-MARK-MASQ chain in iptables/ipvs proxier.
**Which issue(s) this PR fixes**:

Fixes #68447

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by MrHohn** [comment]
@kubernetes/sig-network-pr-reviews 
/ok-to-test

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -396,66 +410,56 @@ func CleanupLeftovers(ipt utiliptables.Interface) (encounteredError bool) {
...
-		writeLine(filterChains, "*filter")
-		for _, chain := range []utiliptables.Chain{kubeServicesChain, kubeExternalServicesChain, kubeForwardChain} {
-			if _, found := existingFilterChains[chain]; found {
-				chainString := string(chain)
-				writeBytesLine(filterChains, existingFilterChains[chain])
-				writeLine(filterRules, "-X", chainString)
+		for _, chain := range append(chains, kubeServicesChain, kubeNodePortsChain, kubePostroutingChain, KubeMarkMasqChain) {
+			if err := ipt.FlushChain(utiliptables.TableNAT, chain); err != nil {
+				if !utiliptables.IsNotFoundError(err) {
+					klog.Errorf("Error delete iptables chain %s: %v", chain, err)
```
Comment (renan) [inline pkg/proxy/iptables/proxier.go]:
```suggestion
					klog.Errorf("Error flush iptables chain %s: %v", chain, err)
```

**Comment by renan** [comment]
Not being able to clean up causes problems with my CNI (calico).

I only had one small suggestion but other than that seems good to me.

**Comment by andrewsykim** [comment]
Should also address https://github.com/kubernetes/kubernetes/issues/80977

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -378,13 +378,27 @@ var iptablesJumpChains = []iptablesJumpChain{
 	{utiliptables.TableNAT, kubePostroutingChain, utiliptables.ChainPostrouting, "kubernetes postrouting rules", nil},
 }
 
-var iptablesCleanupOnlyChains = []iptablesJumpChain{}
+func getExistingChains(ipt utiliptables.Interface, table utiliptables.Table) (chains []utiliptables.Chain, err error) {
+	iptablesData := bytes.NewBuffer(nil)
+	if err = ipt.SaveInto(table, iptablesData); err != nil {
+		klog.Errorf("Failed to execute iptables-save for %s: %v", table, err)
+	} else {
```
Comment (andrewsykim) [inline pkg/proxy/iptables/proxier.go]: return early if err != nil and remove this else?

File: pkg/proxy/iptables/proxier.go (hunk 1 above)
Comment (andrewsykim) [reply pkg/proxy/iptables/proxier.go]: nit: Error flushing

File: cmd/kube-proxy/app/server.go [hunk 3]
```diff
@@ -659,6 +659,8 @@ func getConntrackMax(config kubeproxyconfig.KubeProxyConntrackConfiguration) (in
 // CleanupAndExit remove iptables rules and exit if success return nil
 func (s *ProxyServer) CleanupAndExit() error {
 	encounteredError := userspace.CleanupLeftovers(s.IptInterface)
+	encounteredError = iptables.FlushIptablesChains(s.IptInterface) || encounteredError
+	encounteredError = ipvs.FlushIptablesChains(s.IptInterface) || encounteredError
```
Comment (andrewsykim) [inline cmd/kube-proxy/app/server.go]: What's the reason for calling `FlushIptablesChains` separate from `CleanupLeftovers`? Would `iptables.CleanupLeftovers` error if we don't flush the IPVS chains first? IIRC they share the same chains.

File: pkg/proxy/iptables/proxier.go [hunk 4]
```diff
@@ -396,66 +410,56 @@ func CleanupLeftovers(ipt utiliptables.Interface) (encounteredError bool) {
...
-		// Start with chains we know we need to remove.
-		for _, chain := range []utiliptables.Chain{kubeServicesChain, kubeNodePortsChain, kubePostroutingChain, KubeMarkMasqChain} {
-			if _, found := existingNATChains[chain]; found {
-				chainString := string(chain)
-				writeBytesLine(natChains, existingNATChains[chain]) // flush
-				writeLine(natRules, "-X", chainString)              // delete
+		for _, chain := range append(chains, kubeServicesChain, kubeNodePortsChain, kubePostroutingChain, KubeMarkMasqChain) {
+			if err := ipt.FlushChain(utiliptables.TableNAT, chain); err != nil {
+				if !utiliptables.IsNotFoundError(err) {
+					klog.Errorf("Error flush iptables chain %s: %v", chain, err)
```
Comment (andrewsykim) [inline pkg/proxy/iptables/proxier.go]: nit: "Error flushing ... "

File: pkg/proxy/iptables/proxier.go [hunk 5]
```diff
@@ -378,13 +378,27 @@ var iptablesJumpChains = []iptablesJumpChain{
...
-var iptablesCleanupOnlyChains = []iptablesJumpChain{}
+func getExistingChains(ipt utiliptables.Interface, table utiliptables.Table) (chains []utiliptables.Chain, err error) {
+	iptablesData := bytes.NewBuffer(nil)
+	if err = ipt.SaveInto(table, iptablesData); err != nil {
+		klog.Errorf("Failed to execute iptables-save for %s: %v", table, err)
+	} else {
+		existingNATChains := utiliptables.GetChainLines(table, iptablesData.Bytes())
+		for chain := range existingNATChains {
+			chainString := string(chain)
+			if strings.HasPrefix(chainString, "KUBE-SVC-") || strings.HasPrefix(chainString, "KUBE-SEP-") || strings.HasPrefix(chainString, "KUBE-FW-") || strings.HasPrefix(chainString, "KUBE-XLB-") {
```
Comment (andrewsykim) [inline pkg/proxy/iptables/proxier.go]: 👍 - not deleting `KUBE-SEP-*` chains from iptables proxy was causing all sorts of issues on IPVS clean up.

**Comment by andrewsykim** [comment]
/priority important-soon 

Not cleaning up `KUBE-SEP*` rules also causes issues in the IPVS proxier because it tries to delete a target (like `KUBE-MARK-MASQ`) that is still in use by all the `KUBE-SEP*` chains which results in something like:
```
Error removing iptables rules in ipvs proxier: error deleting chain "KUBE-MARK-MASQ": exit status 1: iptables: Too many links
```

File: cmd/kube-proxy/app/server.go (hunk 3 above)
Comment (liuxu623) [reply cmd/kube-proxy/app/server.go]: Because `iptables.CleanupLeftovers` want delete `KUBE-MARK-MASQ`, but `KUBE-LOAD-BALANCER` and `KUBE-NODE-PORT` link `KUBE-MARK-MASQ`, we shoud flush `KUBE-LOAD-BALANCER` and `KUBE-NODE-PORT` before delete `KUBE-MARK-MASQ`, `KUBE-LOAD-BALANCER` and `KUBE-NODE-PORT` is IPVS chains.

File: pkg/proxy/iptables/proxier.go (hunk 1 above)
Comment (liuxu623) [reply pkg/proxy/iptables/proxier.go]: ok

File: pkg/proxy/iptables/proxier.go (hunk 2 above)
Comment (liuxu623) [reply pkg/proxy/iptables/proxier.go]: ok

File: pkg/proxy/iptables/proxier.go (hunk 4 above)
Comment (liuxu623) [reply pkg/proxy/iptables/proxier.go]: ok

File: pkg/proxy/iptables/proxier.go [hunk 6]
```diff
@@ -396,66 +411,56 @@ func CleanupLeftovers(ipt utiliptables.Interface) (encounteredError bool) {
...
-		filterChains := bytes.NewBuffer(nil)
-		filterRules := bytes.NewBuffer(nil)
-		writeLine(filterChains, "*filter")
-		for _, chain := range []utiliptables.Chain{kubeServicesChain, kubeExternalServicesChain, kubeForwardChain} {
-			if _, found := existingFilterChains[chain]; found {
-				chainString := string(chain)
-				writeBytesLine(filterChains, existingFilterChains[chain])
-				writeLine(filterRules, "-X", chainString)
+		for _, chain := range chains {
+			if err := ipt.FlushChain(utiliptables.TableNAT, chain); err != nil {
```
Comment (andrewsykim) [inline pkg/proxy/iptables/proxier.go]: Looks like we're calling this twice now. Can we rename `CleanupLeftovers` to `DeleteIptablesChains` and have it only do delete?

Comment (liuxu623) [reply pkg/proxy/iptables/proxier.go]: ok

File: cmd/kube-proxy/app/server.go [hunk 7]
```diff
@@ -653,7 +653,9 @@ func getConntrackMax(config kubeproxyconfig.KubeProxyConntrackConfiguration) (in
 // CleanupAndExit remove iptables rules and exit if success return nil
 func (s *ProxyServer) CleanupAndExit() error {
 	encounteredError := userspace.CleanupLeftovers(s.IptInterface)
-	encounteredError = iptables.CleanupLeftovers(s.IptInterface) || encounteredError
+	encounteredError = iptables.FlushIptablesChains(s.IptInterface) || encounteredError
+	encounteredError = ipvs.FlushIptablesChains(s.IptInterface) || encounteredError
+	encounteredError = iptables.DeleteIptablesChains(s.IptInterface) || encounteredError
 	encounteredError = ipvs.CleanupLeftovers(s.IpvsInterface, s.IptInterface, s.IpsetInterface, s.CleanupIPVS) || encounteredError
```
Comment (andrewsykim) [inline cmd/kube-proxy/app/server.go]: Rename ipvs.CleanupLeftovers also to be consistent?

Comment (liuxu623) [reply cmd/kube-proxy/app/server.go]: ipvs.CleanupLeftovers clean ipvs rules, dummy interface, iptables chains and ipset, so we should separate ipvs.FlushIptablesChains form ipvs.CleanupLeftovers ?

Comment (liuxu623) [reply cmd/kube-proxy/app/server.go]: or ipvs.CleanupLeftovers -> ipvs.FluashIPVS + ipvs.DeleteDummyDevice + ipvs. DeleteIptablesChains + ipvs.DeleteIPSet ?

Comment (andrewsykim) [reply cmd/kube-proxy/app/server.go]: ah I see.. let's leave that as is then but can we rename `iptables.DeleteIptablesChains` back to `iptables.CleanupLeftovers`? Just keeps things consistent and easier to follow: "for every proxy we do "Flush" then "CleanupLeftovers"

Comment (liuxu623) [reply cmd/kube-proxy/app/server.go]: ok

File: pkg/proxy/iptables/proxier.go [hunk 8]
```diff
@@ -378,13 +378,28 @@ var iptablesJumpChains = []iptablesJumpChain{
 	{utiliptables.TableNAT, kubePostroutingChain, utiliptables.ChainPostrouting, "kubernetes postrouting rules", nil},
 }
 
-var iptablesCleanupOnlyChains = []iptablesJumpChain{}
+func getExistingKubeNATChains(ipt utiliptables.Interface) (chains []utiliptables.Chain, err error) {
```
[the remark this answers is not in the document]
Comment (andrewsykim) [reply pkg/proxy/iptables/proxier.go]: Maybe for a follow-up, but would be good if there were unit tests for this.

File: pkg/proxy/ipvs/proxier.go [hunk 9]
```diff
@@ -108,6 +108,7 @@ var iptablesChains = []struct {
 	{utiliptables.TableNAT, KubeNodePortChain},
 	{utiliptables.TableNAT, KubeLoadBalancerChain},
 	{utiliptables.TableNAT, KubeMarkMasqChain},
+	{utiliptables.TableNAT, KubeMarkDropChain},
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]: Can we put this change in a separate commit - "flush and delete KUBE-MARK-DROP chain in IPVS clean up"

**Review by andrewsykim — COMMENTED** [review]
/lgtm 
/assign @thockin 

Also Fixes https://github.com/kubernetes/kubernetes/issues/80977

**Comment by thockin** [comment]
I think we have mostly given up on being able to switch modes without a node reboot.  I don't think we should go to heroic lengths to try to make that work automatically.

**Comment by thockin** [comment]
I am very confused.

Why are we trying to delete MARK-MASQ or MARK-DROP?  I feel like I have asked this elsewhere, but I can't find it...

File: pkg/proxy/ipvs/proxier.go (hunk 9 above)
Comment (liuxu623) [reply pkg/proxy/ipvs/proxier.go]: ok

File: pkg/proxy/iptables/proxier.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (liuxu623) [reply pkg/proxy/iptables/proxier.go]: ok

**Comment by andrewsykim** [comment]
Sorry for not adding more context @thockin. Here's what I think is happening based on comments in https://github.com/kubernetes/kubernetes/issues/68447. 

Because IPVS proxy adds the following rule:
```
-t nat -A KUBE-LOAD-BALANCER -j KUBE-MARK-MASQ
```
and `kube-proxy --cleanup` always attempts to clean up rules from all proxy modes (iptables, then ipvs, then userspace), the `iptables` clean up fails with the following:
```
E0905 20:09:11.733419       1 proxier.go:423] Failed to execute iptables-restore for nat: exit status 1 (iptables-restore: line 7 failed
)
```
with line 7 being `-X KUBE-MARK-MASQ` which is still linked by `KUBE-LOAD-BALANCER` (not known to the iptables cleanup). This is a side effect of `kube-proxy --cleanup` cleaning up rules in all modes regardless of the proxy mode + IPVS using the same chain names as the iptables proxy which happens on IPVS mode regardless if we switched from iptables or not.

**Comment by andrewsykim** [comment]
This PR changes the clean up ordering to be "flush all chains in all modes, then delete all chains in all modes" so that unlinked chains can't block delete from another mode.

**Comment by andrewsykim** [comment]
@liuxu623 spoke with @thockin on slack and decided it would be cleaner to just have kube-proxy ignore `KUBE-MARK-MASQ` entirely during clean up because kubelet also manages it. Can we simplify the PR so that it ignores `KUBE-MARK-MASQ` going forward?

**Comment by thockin** [comment]
I hope such a change would eliminate the errors *and* be correct :)

**Comment by liuxu623** [comment]
Why kubelet also manages some iptables chains? I found there are some duplicate code in kubelet_network_linux.go and iptables/proxier.go ipvs/proxier.go.

**Comment by andrewsykim** [comment]
Ran some initial tests based on this PR.

**Before** this PR, the IPVS proxy consistently errors with the following:
```
kube-proxy --cleanup
W0815 18:54:34.588407   35338 server.go:216] WARNING: all flags other than --config, --write-config-to, and --cleanup are deprecated. Please begin using a config file ASAP.
E0815 18:54:34.692430   35338 proxier.go:431] Failed to execute iptables-restore for nat: exit status 1 (iptables-restore: line 22 failed
)
E0815 18:54:34.807099   35338 proxier.go:603] Error removing iptables rules in ipvs proxier: error deleting chain "KUBE-MARK-MASQ": exit status 1: iptables: Too many links.
```
Both the iptables and ipvs clean up fails at `KUBE-MARK-MASQ` because of the following rule:
```
-t nat -A KUBE-LOAD-BALANCER -j KUBE-MARK-MASQ
``` 

If I manually delete that rule and then run clean up, I see no errors. 

**After** the diff, I let kube-proxy do a full resync - which adds the `KUBE-LOAD-BALANCER` chain back and clean up is successful:
```
$ kube-proxy --cleanup
W0815 20:52:08.898254     474 server.go:205] WARNING: all flags other than --config, --write-config-to, and --cleanup are deprecated. Please begin using a config file ASAP.
```

I think KUBE-MARK-MASQ is still being cleaned by the ipvs proxy though, as part of `cleanupIptablesLeftovers` - that probably shouldn't delete the `KUBE-MARK-MASQ` chain as well?

**Comment by liuxu623** [comment]
I think we shouldn't manage KUBE-MARK-MASQ chain and rule in iptables/ipvs proxier, it only should manage by kubelet like KUBE-MARK-DROP.

File: pkg/proxy/ipvs/proxier.go [hunk 10]
```diff
@@ -1553,7 +1552,7 @@ func (proxier *Proxier) createAndLinkeKubeChain() {
 			if chain, ok := existingNATChains[ch.chain]; ok {
 				writeBytesLine(proxier.natChains, chain)
 			} else {
-				writeLine(proxier.natChains, utiliptables.MakeChainLine(kubePostroutingChain))
```
Comment (liuxu623) [inline pkg/proxy/ipvs/proxier.go]: Compare code with iptables proxier, I think there is a mistake.

File: pkg/proxy/ipvs/proxier.go [hunk 11]
```diff
@@ -1553,7 +1552,7 @@ func (proxier *Proxier) createAndLinkeKubeChain() {
...
+				writeLine(proxier.natChains, utiliptables.MakeChainLine(ch.chain))
```
Comment (liuxu623) [inline pkg/proxy/ipvs/proxier.go]: Compare code with iptables proxier, I think there is a mistake.

File: pkg/proxy/ipvs/proxier.go [hunk 12]
```diff
@@ -1550,7 +1550,7 @@ func (proxier *Proxier) createAndLinkeKubeChain() {
...
```
Comment (liuxu623) [inline pkg/proxy/ipvs/proxier.go]: Compare code with iptables proxier, I think there is a mistake.

File: pkg/proxy/ipvs/proxier.go [hunk 13]
```diff
@@ -1553,7 +1550,7 @@ func (proxier *Proxier) createAndLinkeKubeChain() {
...
```
Comment (liuxu623) [inline pkg/proxy/ipvs/proxier.go]: compare code with iptables proxier, I think there is a mistake

File: pkg/proxy/iptables/proxier.go [hunk 14]
```diff
@@ -763,32 +759,14 @@ func (proxier *Proxier) syncProxyRules() {
 			writeLine(proxier.filterChains, utiliptables.MakeChainLine(chainName))
 		}
 	}
-	for _, chainName := range []utiliptables.Chain{kubeServicesChain, kubeNodePortsChain, kubePostroutingChain, KubeMarkMasqChain} {
```
Comment (andrewsykim) [inline pkg/proxy/iptables/proxier.go]: Confused, why we are removing `KUBE-POSTROUTING` here now?

File: pkg/proxy/iptables/proxier.go [hunk 15]
```diff
@@ -763,32 +759,14 @@ func (proxier *Proxier) syncProxyRules() {
...
 			writeBytesLine(proxier.natChains, chain)
 		} else {
 			writeLine(proxier.natChains, utiliptables.MakeChainLine(chainName))
 		}
 	}
 
-	// Install the kubernetes-specific postrouting rules. We use a whole chain for
-	// this so that it is easier to flush and change, for example if the mark
-	// value should ever change.
-	writeLine(proxier.natRules, []string{
```
Comment (andrewsykim) [inline pkg/proxy/iptables/proxier.go]: We still need this as far as I can tell, where is this rule getting replaced in the iptables proxy?

Comment (andrewsykim) [reply pkg/proxy/iptables/proxier.go]: (We still need to use `KUBE-MARK-MASQ` in the proxy, we just shouldn't try to flush/delete the chain as part of `--cleanup`)

File: pkg/proxy/ipvs/proxier.go (hunk 13 above)
Comment (andrewsykim) [reply pkg/proxy/ipvs/proxier.go]: In this case `ch.chain` is not in `existingNATChains` so we fallback to `KUBE-POSTROUTING`. Doesn't seem like a bug to me (yet), but can you fix this in a separate PR please?

File: pkg/proxy/ipvs/proxier.go [hunk 16]
```diff
@@ -1571,23 +1568,6 @@ func (proxier *Proxier) createAndLinkeKubeChain() {
 		}
 	}
 
-	// Install the kubernetes-specific postrouting rules. We use a whole chain for
-	// this so that it is easier to flush and change, for example if the mark
-	// value should ever change.
-	writeLine(proxier.natRules, []string{
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]: Same here - we still need these rules, where are they getting replaced?

File: pkg/proxy/iptables/proxier.go (hunk 15 above)
Comment (liuxu623) [reply pkg/proxy/iptables/proxier.go]: you can see [kubelet_network_linux.go#L99-L104](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_network_linux.go#L99-L104) and [kubelet_network_linux.go#L90-L93](
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_network_linux.go#L90-L93), kubelet will create rules, it just like `KUBE-MARK-DROP`, iptables/ipvs also need `KUBE-MARK-DROP` rule, but it only create by kubelet at [kubelet_network_linux.go#L56-L59](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_network_linux.go#L56-L59), iptables/ipvs don't manage `KUBE-MARK-DROP`.

File: pkg/kubelet/kubelet_network.go [hunk 17]
```diff
@@ -27,7 +27,6 @@ import (
 
 const (
 	// KubeMarkMasqChain is the mark-for-masquerade chain
-	// TODO: clean up this logic in kube-proxy
```
Comment (liuxu623) [inline pkg/kubelet/kubelet_network.go]: @andrewsykim I just finished this TODO

File: pkg/proxy/ipvs/proxier.go (hunk 13 above)
Comment (liuxu623) [reply pkg/proxy/ipvs/proxier.go]: ok, I will fix this in another PR.

File: pkg/proxy/iptables/proxier.go (hunk 15 above)
Comment (liuxu623) [reply pkg/proxy/iptables/proxier.go]: not only clean up, we also shouldn't create `KUBE-MARK-MASQ` `KUBE-POSTROUTING` chains and rules in iptables/ipvs proxier.

Comment (andrewsykim) [reply pkg/proxy/iptables/proxier.go]: Can we open a separate issue for this and address in a separate PR please? Ideally this PR only fixes issues reported in #68447.

Comment (liuxu623) [reply pkg/proxy/iptables/proxier.go]: ok

**Comment by andrewsykim** [comment]
LGTM, thanks for iterating @liuxu623 

@thockin can we get another review please?

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Comment by liuxu623** [comment]
@andrewsykim @thockin Thank you for review my code !
